city_metrix/layers/ndvi_sentinel2_gee.py

```diff
@@ -0,0 +1,65 @@
+import ee
+from tools.xarray_tools import convert_ratio_to_percentage
+from .layer import Layer, get_image_collection
+
+class NdviSentinel2(Layer):
+    """"
+    NDVI = Sentinel-2 Normalized Difference Vegetation Index
+    param: year: The satellite imaging year.
+    return: a rioxarray-format DataArray
+    Author of associated Jupyter notebook: Ted.Wong@wri.org
+    Notebook: https://github.com/wri/cities-cities4forests-indicators/blob/dev-eric/scripts/extract-VegetationCover.ipynb
+    Reference: https://en.wikipedia.org/wiki/Normalized_difference_vegetation_index
+    """
+    def __init__(self, year=None, **kwargs):
+        super().__init__(**kwargs)
+        self.year = year
+
+    def get_data(self, bbox):
+        if self.year is None:
+            raise Exception('NdviSentinel2.get_data() requires a year value')
+
+        start_date = "%s-01-01" % self.year
+        end_date = "%s-12-31" % self.year
+
+        # Compute NDVI for each image
+        def calculate_ndvi(image):
+            ndvi = (image
+                    .normalizedDifference(['B8', 'B4'])
+                    .rename('NDVI'))
+            return image.addBands(ndvi)
+
+        s2 = ee.ImageCollection("COPERNICUS/S2_HARMONIZED")
+        ndvi = (s2
+                .filterBounds(ee.Geometry.BBox(*bbox))
+                .filterDate(start_date, end_date)
+                .map(calculate_ndvi)
+                .select('NDVI')
+                 )
+
+        ndvi_mosaic = ndvi.qualityMosaic('NDVI')
+
+        ic = ee.ImageCollection(ndvi_mosaic)
+        ndvi_data = get_image_collection(ic, bbox, 10, "NDVI")
+
+        xdata = ndvi_data.to_dataarray()
+
+        return xdata
+
+    def prepare_for_map_rendering(self, data, ndvi_threshold=0.4, convert_to_percentage=True, **kwargs):
+        """
+        Applies the standard post-processing adjustment used for rendering of NDVI including masking
+        to a threshold and conversion to percentage values.
+        :param ndvi_threshold: (float) minimum threshold for keeping values
+        :param convert_to_percentage: (bool) controls whether NDVI values are converted to a percentage
+        :return: A rioxarray-format DataArray
+        """
+        # Remove values less than the specified threshold
+        if ndvi_threshold is not None:
+            data = data.where(data >= ndvi_threshold)
+
+        # Convert to percentage in byte data_type
+        if convert_to_percentage is True:
```

Comment:
I'm not sure if this was so the map shows percentages instead of decimals, or just to reduce the size of the output from float to byte, since the current frontend just directly loads GeoTiffs. Since they're remaking the frontend to actually use map tiles, I'm not sure if this is necessary anymore, we can confirm with Ted and Saif

Reply:
In this case, I was describing the outcome of the function rather than the intended use. No matter what, I personally like reducing some of the data variability by rounding since I assume the values are not too precise. Having smaller data volume is a big advantage too.

@tedw0ng - What's your opinion?

Reply:
It's fine to serve the raw data. 0.4 is not a universal threshold for vegetated, and some users might want to apply their own thresholds. If someone wants to see what happens when the 0.4 threshold is applied, they can look at the indicator.

I don't think it's necessary to provide a function for the secondary processing. Thresholds and percents are things that are easily done by a downstream script or GIS.

Reply:
@tedw0ng @S-AI-F Such a function is useful for testing, so I will move it out of the layers class and into testing.

Reply:
Post-processing method removed from Layers.